crates/bevy_ecs/src/world/component_constants.rs

```diff
@@ -0,0 +1,23 @@
+use super::*;
+use crate as bevy_ecs;
+/// Internal components used by bevy with a fixed component id.
+/// Constants are used to skip [`TypeId`] lookups in hot paths.
+
+/// [`ComponentId`] for [`OnAdd`]
+pub const ON_ADD: ComponentId = ComponentId::new(0);
+/// [`ComponentId`] for [`OnInsert`]
+pub const ON_INSERT: ComponentId = ComponentId::new(1);
+/// [`ComponentId`] for [`OnRemove`]
+pub const ON_REMOVE: ComponentId = ComponentId::new(2);
```

Comment:
I think you would have to pay something for lazy_static since you need to ensure it is actually initialized, not a huge fan of the separation here either but not sure of a suitable alternative.